src/ApplicationCore/Services/BasketService.cs

```diff
@@ -59,11 +59,22 @@ public async Task TransferBasketAsync(string anonymousId, string userName)
         {
             Guard.Against.NullOrEmpty(anonymousId, nameof(anonymousId));
             Guard.Against.NullOrEmpty(userName, nameof(userName));
-            var basketSpec = new BasketWithItemsSpecification(anonymousId);
-            var basket = (await _basketRepository.FirstOrDefaultAsync(basketSpec));
-            if (basket == null) return;
-            basket.SetNewBuyerId(userName);
-            await _basketRepository.UpdateAsync(basket);
+            var anonymousBasketSpec = new BasketWithItemsSpecification(anonymousId);
```

Comment:
Can we add some tests here? There should already be a test file for `TransferBasketAsync`, it just needs updated to account for the new logic.

Reply:
I plan to write some tests related to following:

- removes anonymous basket after transfer
- preserves existing basket items in the user basket while transferring anonymous basket items 
- creates a user basket before transferring if not exists

Reply:
Those sound great! You could also write a few unit tests that mock out the `_basketRepository` and ensure all the various methods are invoked the number of times you expect for the different scenarios.

Reply:
```
        [Fact]
        public async Task InvokesBasketRepositoryFirstOrDefaultAsyncOnceIfAnonymousBasketNotExists()
        {
            Basket basket = null;
            _mockBasketRepo.Setup(x => x.FirstOrDefaultAsync(It.IsAny<BasketWithItemsSpecification>())).ReturnsAsync(basket);
            var basketService = new BasketService(_mockBasketRepo.Object, null);
            await basketService.TransferBasketAsync("AnIdThatDoesNotExist", "testuser@microsoft.com");
            _mockBasketRepo.Verify(x => x.FirstOrDefaultAsync(It.IsAny<BasketWithItemsSpecification>()), Times.Once);
        }
```

I might use your opinion on this. Here, I am trying to confirm that it doesn't look for or create a user basket if anonymous basket to transfer from doesn't exist.

Reply:
That is looking good! You could update the verify to check that the specific value was used when you tried to invoke it. So instead of passing in `It.IsAny<BasketWithItemsSpecification>()` you could create the two different specifications and pass them in. That way you can actually have multiple verifies on the `FirstOrDefaultAsync` call.

One that confirms that `FirstOrDefaultAsync` was invoked with the `anonymousBasketSpec`, and another that confirms `FirstOrDefaultAsync` was **never** invoked for the `userBasketSpec`.

Reply:
```csharp
private readonly string _nonexistentAnonymousBasketBuyerId = "nonexistent-anonymous-basket-buyer-id";
private readonly string _existentUserBasketBuyerId = "testuser@microsoft.com";
[Fact]
public async Task InvokesBasketRepositoryFirstOrDefaultAsyncOnceIfAnonymousBasketNotExists()
{
    var nullBasket = null as Basket;
    var userBasket = new Basket(_existentUserBasketBuyerId);
    var anonymousBasketSpec = new BasketWithItemsSpecification(_nonexistentAnonymousBasketBuyerId);
    var userBasketSpec = new BasketWithItemsSpecification(_existentUserBasketBuyerId);
    _mockBasketRepo.Setup(x => x.FirstOrDefaultAsync(anonymousBasketSpec)).ReturnsAsync(nullBasket);
    _mockBasketRepo.Setup(x => x.FirstOrDefaultAsync(userBasketSpec)).ReturnsAsync(userBasket);
    var basketService = new BasketService(_mockBasketRepo.Object, null);
    await basketService.TransferBasketAsync("AnIdThatDoesNotExist", "testuser@microsoft.com");
    _mockBasketRepo.Verify(x => x.FirstOrDefaultAsync(anonymousBasketSpec), Times.Once);
    _mockBasketRepo.Verify(x => x.FirstOrDefaultAsync(userBasketSpec), Times.Never);
}
```

I've just prepared this faulty code to be able to discuss on it. Something like this doesn't work because of 2 main problems:
1. Last setup overrides the first setup as explained on the explanation of the method: If more than one setup is specified for the same method or property, the latest one wins and is the one that will be executed.
2. instead of x => Since BasketWithItemsSpecification is a reference type, it will not match with the specification object created in the TransferBasketAsync service method. If an equality checker existed for the type ISpecification, I thought I could update ReturnsAsync and Verify part so that it will return the right basket for each specification.

Reply:
Ah, yes - I see what you mean. For now, we don't have to do the verify of the specific Specification but let's still do the verification of the number of times the method was called. 

For issue #1 you described above, you can actually mock out a "sequence" like this:

```
            _mockBasketRepo.SetupSequence(x => x.FirstOrDefaultAsync(It.IsAny<BasketWithItemsSpecification>()))
                .ReturnsAsync(anonymousBasket)
                .ReturnsAsync(userBasket);
```

The above snippet will return an `anonymousBasket` on the first time the `FirstOrDefaultAsync` is invoked, and will return the different `userBasket` the second time it's invoked. Hopefully that helps!

Reply:
Thank you @efleming18, your suggestion helped. I will proceed accordingly.

Reply:
https://github.com/dotnet-architecture/eShopOnWeb/pull/408/commits/9fdb60db714df7b15cde768cb7fc5d659043839b